src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -3227,6 +3227,132 @@ tx_gateway_frontend::get_all_transactions() {
+
+          if (!stm) {
+              vlog(
+                txlog.error,
+                "can't get tm stm of the {}' partition",
+                model::tx_manager_ntp);
+              return ss::make_ready_future<result<tm_transaction, tx_errc>>(
+                tx_errc::stm_not_found);
+          }
+
+          auto gate_lock = gate_guard(self._gate);
+          return stm->read_lock().then([&self, stm, tid](
+                                         ss::basic_rwlock<>::holder unit) {
+              return with(
+                       stm,
+                       tid,
+                       "get_tx",
+                       [&self, stm, tid]() { return self.do_get_tx(stm, tid); })
+                .finally([u = std::move(unit)] {});
+          });
+      });
+}
+
+ss::future<result<tm_transaction, tx_errc>> tx_gateway_frontend::do_get_tx(
+  ss::shared_ptr<tm_stm> stm, kafka::transactional_id tid) {
+    auto term_opt = co_await stm->sync();
+    if (!term_opt.has_value()) {
+        if (term_opt.error() == tm_stm::op_status::not_leader) {
+            co_return tx_errc::not_coordinator;
+        }
+        co_return tx_errc::coordinator_not_available;
+    }
+    auto term = term_opt.value();
+
+    auto tx_opt = co_await stm->get_tx(tid);
+    if (!tx_opt.has_value()) {
+        auto status = tx_opt.error();
+        tx_errc err = tx_errc::unknown_server_error;
+        if (status == tm_stm::op_status::not_leader) {
+            err = tx_errc::leader_not_found;
+        } else if (status == tm_stm::op_status::timeout) {
+            err = tx_errc::timeout;
+        } else if (status == tm_stm::op_status::not_found) {
+            err = tx_errc::tx_id_not_found;
+        }
+        co_return err;
+    }
+
+    auto tx = tx_opt.value();
+
+    if (term != tx.etag) {
+        if (!is_fetch_tx_supported()) {
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        auto r1 = co_await fetch_tx(tx.id, tx.etag);
+        if (!r1) {
+            vlog(txlog.warn, "can't find old tx:{} etag:{}", tx.id, tx.etag);
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        auto old_tx = r1.value();
+
+        if (
+          old_tx.status != tm_transaction::tx_status::ongoing
+          && old_tx.status != tm_transaction::tx_status::ready) {
+            vlog(
+              txlog.warn,
+              "fetched tx:{} pid:{} etag:{} has unexpected status {}",
+              tx.id,
+              old_tx.pid,
+              old_tx.etag,
+              old_tx.status);
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        if (old_tx.pid != tx.pid) {
+            vlog(
+              txlog.warn,
+              "fetched tx:{} pid:{} etag:{} should have pid:{}",
+              tx.id,
+              old_tx.pid,
+              old_tx.etag,
+              tx.pid);
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        old_tx.etag = term;
+        auto r2 = co_await stm->update_tx(old_tx, term);
```

Comment:
This method is named `do_get_tx` but here it updates the `tx`  why the update is required, `get` in the method name suggest that the metod doesn't mutate the state of transaction

Reply:
usually it's true but there are exceptions; here we reconcile the state and save it to make sure future requests will see the same